### DISCUSSION TITLE
Where to put calculations on data, in plugin or in main code?

### DISCUSSION THREAD

**Discussion by CarolynBuck-Gengler** [jspsych/jsPsych#894] [Q&A] [answered]
When creating a new plugin, is it better programming practice to compute accuracy (and/or translate key codes into characters typed) in the plugin or in the main code that calls the plugin? (or whatever other data massaging one wants to do along the way.) Doing it in the plugin obviously assumes the plugin knows what extra information is needed, and the particulars of how to compute it, but if it is a specialized plugin, maybe that's the case anyway. 
Or should one always be writing a plugin with an eye toward sharing it, and thus not putting too much special knowledge into it? 
OTOH, perhaps other end users would be able to do their own massaging of the returned data (or even the plugin) if the extra information is not useful, but make use of it if it is useful?
Thanks!

**Comment by jodeleeuw** [comment, marked as answer]
My inclination would be to do any calculations that are universal to use cases of the plugin. So converting key codes to characters seems generally useful (and something we are planning to transition to across the library in the future, see #396). You could compute accuracy if the necessary information is in the plugin (like in [jspsych-categorize-html](https://www.jspsych.org/plugins/jspsych-categorize-html/)). 

I don't think there's too much harm in doing extra computation in the plugin. It potentially makes the data file longer, but methods like `jsPsych.data.get().ignore()` can solve that issue by trimming irrelevant variables. 

The only case where I wouldn't want to put calculation in the plugin is if I had to code in the contextual information directly and didn't turn it into a parameter. That would require the user of the plugin to edit the plugin file directly to use it for it's normal purpose, and that's something we avoid in the design of plugins.
